### PR TITLE
Add fuzzers for bus match parsing code

### PR DESCRIPTION
**Pull request by keszybz**
I wrote this to investigate https://bugzilla.redhat.com/show_bug.cgi?id=1935084.
Some minor issues were found, but nothing that'd explain that bug.

### CODE REVIEW

File: src/libsystemd/sd-bus/bus-match.c [hunk 1]
```diff
@@ -1017,14 +1017,12 @@ const char* bus_match_node_type_to_string(enum bus_match_node_type t, char buf[]
...
 void bus_match_dump(FILE *out, struct bus_match_node *node, unsigned level) {
-        _cleanup_free_ char *pfx = NULL;
         char buf[32];
 
         if (!node)
                 return;
 
-        pfx = strrep("  ", level);
-        fprintf(out, "%s[%s]", strempty(pfx), bus_match_node_type_to_string(node->type, buf, sizeof(buf)));
+        fprintf(out, "%*s[%s]", 2 * level, "", bus_match_node_type_to_string(node->type, buf, sizeof(buf)));
```
Comment (yuwata) [inline src/libsystemd/sd-bus/bus-match.c:1028]: Maybe, a white-space is missing: `""` -> `" "`.

File: src/libsystemd/sd-bus/fuzz-bus-match.c [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+
+                offset = end ? (size_t) (end - (char*) data + 1) : size;
+
+                struct bus_match_component *components;
+                unsigned n_components;
+                r = bus_match_parse(line, &components, &n_components);
+                if (r == -EINVAL) {
+                        log_debug_errno(r, "Failed to parse line: %m");
+                        continue;
+                }
```
Comment (yuwata) [inline src/libsystemd/sd-bus/fuzz-bus-match.c]: `bus_match_parse()` may also return `-ENOMEM`, if `-ENOMEM` is ignored here, then `components` and `n_components` are used without being initialized below.
So, `if (r == -EINVAL)` -> `if (r < 0)`

File: src/libsystemd/sd-bus/bus-match.c [hunk 3]
```diff
@@ -878,9 +878,9 @@ char *bus_match_to_string(struct bus_match_component *components, unsigned n_com
         }
 
         r = fflush_and_check(f);
+        safe_fclose(f);
         if (r < 0)
                 return NULL;
```
Comment (yuwata) [inline src/libsystemd/sd-bus/bus-match.c:886]: Is it not necessary to free `buffer` on failure? How about the following?
```
_cleanup_free_ char *buffer = NULL;
...
return TAKE_PTR(buffer);
```

**Review by yuwata — CHANGES_REQUESTED** [review]

File: src/libsystemd/sd-bus/bus-match.c (hunk 1 above)
Comment (keszybz) [reply src/libsystemd/sd-bus/bus-match.c:1028]: No, it's using output width to generate the indent. So all the output comes from the padding, and the argument itself is just a placeholder.

File: src/libsystemd/sd-bus/fuzz-bus-match.c (hunk 2 above)
Comment (keszybz) [reply src/libsystemd/sd-bus/fuzz-bus-match.c]: True. I wanted to treat EINVAL differently, but I forgot to add the other leg of the conditional.

File: src/libsystemd/sd-bus/bus-match.c (hunk 3 above)
Comment (keszybz) [reply src/libsystemd/sd-bus/bus-match.c:886]: I merged your PR and rebased my changes on top and used the same pattern here.

**Comment by keszybz** [comment]
Updated.

File: src/libsystemd/sd-bus/bus-match.c (hunk 1 above)
Comment (yuwata) [reply src/libsystemd/sd-bus/bus-match.c:1028]: Ah, yes, you are right. Sorry for the noise.

**Comment by yuwata** [comment]
LGTM.

**Comment by yuwata** [comment]
packit failures are not related, and hopefully fixed by #18902.

**Pull request merged**